### PR TITLE
Version 0.1

### PR DESCRIPTION
**Pull request by mecha**
Initial release, containing the abstract implementations.

## Added
- Abstract implementations;
- Functional tests.

### CODE REVIEW

**Comment by mecha** [comment]
Once https://github.com/Dhii/di-interface/pull/1 is merged, this package can depend on `dhii/di-interface : ^0.1`.

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
[This](https://github.com/Dhii/di-abstract/blob/release/0.1/src/AbstractCompositeContainer.php#L91) will not work with `Traversable`. Try [`iterator_to_array()`](http://php.net/manual/en/function.iterator-to-array.php)

**Comment by mecha** [comment]
@XedinUnknown Why won't [it](https://github.com/Dhii/di-abstract/blob/release/0.1/src/AbstractCompositeContainer.php#L91) work with `Traversable`? The array pointer PHP functions are all compatible with `Iterator` implementations, and according to the PHP manual the `Iterator` interface is what makes objects `Traversable`.

> Abstract base interface that cannot be implemented alone. Instead it must be implemented by either IteratorAggregate or Iterator.

**Comment by XedinUnknown** [comment]
@mecha, where does it say that PHP array pointer functions are compatible with `Traversable`? IMHO, this is not possible by design, at least because `Iterator` is not the only `Traversable` interface.

To make things simpler, modify the test to use a `Traversable` instead of an array, and if it passes, I will approve.

**Comment by mecha** [comment]
@XedinUnknown I happen to know from experience that it works. I've also just tested it across all PHP versions using [this tool](https://3v4l.org/) using the code:

```php
function foobar(Traversable $t)
{
    return end($t);
}
echo foobar(new ArrayIterator(array(1, 21, 42)));
```

I couldn't find any documentation that listed what classes (specifically built-in classes) implement `Traversable`. In fact I think nothing implements this, save for `array()` internally.

Still, if you prefer this done differently I will comply.

**Comment by XedinUnknown** [comment]
@mecha, please try [this](https://3v4l.org/fpoIK). Function `barfoo()` should explain why there's no error.

**Comment by mecha** [comment]
@XedinUnknown Updated PR.

Introduced [`AbstractCompositeContainer#_getContainersReversed()`][1]. This can be overridden by implementations if an iterator is needed.

[`AbstractCompositeContainer#_hasDelegated()`][2] now relies on that method and iterates on whatever that method returns.

[1]: https://github.com/Dhii/di-abstract/blob/release/0.1/src/AbstractCompositeContainer.php#L130
[2]: https://github.com/Dhii/di-abstract/blob/release/0.1/src/AbstractCompositeContainer.php#L90

**Review by XedinUnknown — APPROVED** [review]

**Comment by XedinUnknown** [comment]
@mecha, last look before release please.

File: src/AbstractCompositeContainer.php [hunk 1]
```diff
@@ -42,7 +42,7 @@ protected function _add(BaseContainerInterface $container)
      *
      * @todo To check if the $container instance should play a part in ID generation.
```
Comment (mecha) [inline src/AbstractCompositeContainer.php]: @XedinUnknown Remove this tag?

**Comment by mecha** [comment]
@XedinUnknown There is also [this `todo` tag](https://github.com/Dhii/di-abstract/blob/master/src/AbstractContainer.php#L62). Worth discussing as I can't see any scenario where the method can return `null`, seeing as how [`_resolveDefinition()`](https://github.com/Dhii/di-abstract/blob/master/src/AbstractContainer.php#L252) either returns or throws an exception.

**Comment by XedinUnknown** [comment]
@mecha, check now please.

File: src/AbstractCompositeContainer.php [hunk 2]
```diff
@@ -62,11 +63,15 @@ protected function _createContainerId(BaseContainerInterface $container)
...
+     * @throws NotFoundException If none of the inner containers have a matching service.
      */
     protected function _getDelegated($id)
     {
-        $having = $this->_hasDelegated($id);
+        if (!($having = $this->_hasDelegated($id))) {
+            throw $this->_createNotFoundException(sprintf('Could not create service for ID "%1$s": no service defined', $id));
+        }
 
         return ($having)
```
Comment (mecha) [inline src/AbstractCompositeContainer.php]: This return ternary operator is now obsolete. I would also recommend inverting the `if` condition for cleaner code.

```php
if ($having = $this->_getDelegated($id)) {
    return $having->get($id)
}

throw $this->_createNotFoundException(sprintf('Could not create service for ID "%1$s": no service defined', $id));
```

Comment (XedinUnknown) [reply src/AbstractCompositeContainer.php]: Fixed check. I think I prefer to throw in the condition, though.

Comment (mecha) [reply src/AbstractCompositeContainer.php]: Up to you. Both look good to me now after the ternary check fix.

**Pull request merged**